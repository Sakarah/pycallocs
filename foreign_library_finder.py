import allocs
import importlib
import importlib.abc
import importlib.machinery
import sys

# These can be modified by the user to change the finder search paths
lib_extension = ".so"

# The blank search path has a special meaning as dlopen is using its own search 
# paths when the file name does not contain any /
search_paths = [""]

class ForeignLibraryFinder(importlib.abc.MetaPathFinder):
    """
        Meta path finder searching foreign libraries usable
        with allocs.ForeignLibraryLoader.
    """

    def find_spec(fullname, path, target):
        if path is not None:
            # Should we support submodules ?
            return None
        
        for base_path in search_paths:
            try:
                filename = base_path + fullname + lib_extension
                loader = allocs.ForeignLibraryLoader(filename)
                return importlib.machinery.ModuleSpec(fullname, loader, origin=filename)
            except ImportError:
                continue

        return None

sys.meta_path.append(ForeignLibraryFinder)
