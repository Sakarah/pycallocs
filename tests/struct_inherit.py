import foreign_library_finder
foreign_library_finder.search_paths.append("libs/")
import inheritance as m

b = m.base(0)
m.print_base(b)
m.dynamic_print(b)

d = m.derivated((1,), 0)
m.print_base(d)
m.print_derivated(d)
m.dynamic_print(d)

l = m.leaf(((2,), 3.14), 'l')
m.print_base(l)
m.print_derivated(l)
m.print_leaf(l)
m.dynamic_print(l)

m.dynamic_print(m.int_32.array([3]))
