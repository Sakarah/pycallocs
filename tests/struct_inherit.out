Base: 0
Base: 0
Base: 1
Derivated: 1, 0.000000
Derivated: 1, 0.000000
Base: 2
Derivated: 2, 3.140000
Leaf: 2, 3.140000, l
Leaf: 2, 3.140000, l
Unrecognized object
